src/monodroid/jni/monodroid-glue.c

```diff
@@ -1420,6 +1418,77 @@ java_gc (JNIEnv *env)
 }
 
 static void
+set_bridge_processing_field (MonoVTable *jnienv_vtable, MonoClassField *android_runtime_jnienv_bridge_processing_field, mono_bool value)
+{
+	mono.mono_field_static_set_value (jnienv_vtable, android_runtime_jnienv_bridge_processing_field, &value);
+}
+
+struct MonoDroidDomainId {
+	int domain_id;
+	MonoClass *android_runtime_jnienv;
+	MonoClassField *android_runtime_jnienv_bridge_processing_field;
+	MonoVTable *jnienv_vtable;
+
+	struct MonoDroidDomainId* next;
+};
+
+typedef struct MonoDroidDomainId MonoDroidDomainId;
+MonoDroidDomainId *domains_list;
+
+static void
+add_monodroid_domain (MonoDomain *domain)
+{
+	MonoDroidDomainId *node = calloc (1, sizeof (MonoDroidDomainId));
+
+	/* We need to prefetch all these information prior to using them in gc_cross_reference as all those functions
+	 * use GC API to allocate memory and thus can't be called from within the GC callback as it causes a deadlock
+	 * (the routine allocating the memory waits for the GC round to complete first)
+	 */
+	MonoClass *jnienv = monodroid_get_class_from_name (&mono, domain, "Mono.Android", "Android.Runtime", "JNIEnv");;
+	node->domain_id = mono.mono_domain_get_id (domain);
+	node->android_runtime_jnienv = jnienv;
+	node->android_runtime_jnienv_bridge_processing_field = mono.mono_class_get_field_from_name (jnienv, "BridgeProcessing");
+	node->jnienv_vtable = mono.mono_class_vtable (domain, jnienv);
+	node->next = domains_list;
+
+	domains_list = node;
+}
+
+static void
+remove_monodroid_domain (int domain_id)
```

Comment:
Why use `int domain_id` instead of `MonoDomain *domain`? The only caller of `remove_monodroid_domain()` has access to the `MonoDomain*` value.

Additionally, you said it should now be "closer in spirit to Java.Interop." Instead of a struct containing `MonoClass*`, `MonoClassField*`, `MonoVTable*`, and `int domain_id` (or `MonoDomain*`), [Java.interop keeps two parallel arrays](https://github.com/xamarin/java.interop/blob/3256144/src/java-interop/java-interop-gc-bridge-mono.c#L38-L40):

	int                                 BridgeProcessing_vtables_count, BridgeProcessing_vtables_length;
	MonoDomain                        **BridgeProcessing_domains;
	MonoVTable                        **BridgeProcessing_vtables;

I'm not opposed to having a structure, but I don't see why you need the other fields, as opposed to:

    struct MonodroidBridgeProcessingInfo {
        MonoDomain                           *domain;
        MonoVTable                           *vtable;
        struct MonodroidBridgeProcessingInfo *next;
    };